### PR TITLE
Limit the use of regexes in the RFC 7235 challenge parser.

### PR DESCRIPTION
**Pull request by swankjesse**
Creating patterns dynamically to do comma-escaping is pretty awkward.
This is more code but I have more confidence in it.

### CODE REVIEW

File: okhttp/src/main/java/okhttp3/internal/http/HttpHeaders.java [hunk 1]
```diff
@@ -179,104 +145,168 @@ public static Headers varyHeaders(Headers requestHeaders, Headers responseHeader
...
+    while (!buffer.exhausted() && buffer.getByte(0) == b) {
+      count++;
+      buffer.readByte();
+    }
+    return count;
+  }
+
+  /**
+   * Reads a double-quoted string, unescaping quoted pairs like {@code \"} to the 2nd character in
+   * each sequence. Returns the unescaped string, or null if the buffer doesn't isn't prefixed with
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/internal/http/HttpHeaders.java]: I don't like no double negatives

**Comment by Vampire** [comment]
@swankjesse, @yschimke 
Are you aware that the tests only pass because one was ignored and for four the expectation was rewritten?
The current code does not do what the JavaDoc of the method states.
The JavaDoc of `okhttp3.Response#challenges` states `Invalid headers and challenges are ignored.`, while now some invalid things are accepted (the four tests with changed expectations).

**Comment by swankjesse** [comment]
Yeah, I think it's a net improvement. We could definitely reject the entire line if any challenge is malformed, but I don't think that's necessary.

**Comment by Vampire** [comment]
Well, I think my reasoning was, if a line does not match my overall regex, it cannot be determined reliably where to split and where the error really is, so it would be better to discard the line as documented in the JavaDoc.

You now even accept invalid challenges that have a token68 and auth params set.
But as the cases you changed are mostly edge cases of faulty implementations, I guess it is ok to have it like that.
Just wanted to make sure it is deliberate :-)

**Comment by yschimke** [comment]
@Vampire Yes - I'm ok with that, see my reasoning here https://github.com/square/okhttp/pull/4252#issuecomment-419799502

**Comment by Vampire** [comment]
Oh, sure, just wanted to make sure. :-)

**Pull request merged**